dipy/reconst/dki.py

```diff
@@ -700,8 +800,427 @@ def split_dki_param(dki_params):
+
+    .. [Raf2015] Neto Henriques, R., Correia, M.M., Nunes, R.G., Ferreira,
+        H.A. (2015). Exploring the 3D geometry of the diffusion kurtosis
+        tensor - Impact on the development of robust tractography procedures
+        and novel biomarkers. NeuroImage 111: 85-99.
+        doi:10.1016/j.neuroimage.2015.02.004
+    """
+    # flat data
+    outshape = dki_params.shape[:-1]
+    dki_params = dki_params.reshape((-1, dki_params.shape[-1]))
+
+    # Split data
+    evals, evecs, kt = split_dki_param(dki_params)
+
+    # Initialize AKC matrix
+    V = sphere.vertices
+    kODF = np.zeros((len(kt), len(V)))
+
+    # select non-zero voxels
+    # (when #677 is merged replace this code lines by function _positive_evals)
+    # rel_i = _positive_evals(evals[..., 0], evals[..., 1], evals[..., 2])    
+    er = np.finfo(evals.ravel()[0]).eps * 1e3
+    rel_i = np.logical_and(evals[..., 0] > er,
+                           np.logical_and(evals[..., 1] > er,
+                                          evals[..., 2] > er))
+    kt = kt[rel_i]
+    evecs = evecs[rel_i]
+    evals = evals[rel_i]
+    kODFi = kODF[rel_i]
+
+    # loop over all relevant voxels
+    for vox in range(len(kt)):
+        # Compute dimensionless tensor U
+        R = evecs[vox]
+        dt = np.dot(np.dot(R, np.diag(evals[vox])), R.T)
+        MD = (dt[0, 0] + dt[1, 1] + dt[2, 2]) / 3
+        U = np.linalg.pinv(dt) * MD
+
+        # loop over all directions
+        sampledODF = np.zeros(len(V))
+        for i in range(len(V)):
+             sampledODF[i] = _dki_odf_core(V[i], kt[vox], U, alpha)
+
+        kODFi[vox] = sampledODF
+
+    # reshape data according to input data
+    kODF[rel_i] = kODFi
+    kODF = kODF.reshape((outshape + (len(V),)))
+    return kODF
+
+
+def _dki_odf_core(n, kt, U, alpha=4):
+    """ Helper function thar compute the DKI based ODF estimation of a voxel
+    along the given directions n
+    
+    Parameters
+    ----------
+    n : array (3,)
+        arrays containing the three cartesian coordinates of direction n
+    kt : array (15,)
+        elements of the kurtosis tensor of the voxel.
+    U : array (3, 3)
+        dimensionless tensor defined by $U = \overline{D} \times DT^{-1}$,
+        where \overline{D} is the mean diffusivity and DT the diffusion tensor
+    alpha : float, optional
+        Radial weighting power. Default is 4.
+    
+    Returns
+    -------
+    dki_odf : float
+        The DKI-ODF value for direction n.
+    """
+    # Compute elements of matrix V
+    Un = np.dot(U, n)
+    nUn = np.dot(n, Un)
+    V00 = Un[0]**2 / nUn
+    V11 = Un[1]**2 / nUn
+    V22 = Un[2]**2 / nUn
+    V01 = Un[0]*Un[1] / nUn
+    V02 = Un[0]*Un[2] / nUn
+    V12 = Un[1]*Un[2] / nUn
+
+    # diffusion ODF
+    ODFg = (1./nUn) ** ((alpha+1.)/2.)
+
+    # Estimate ODF
+    ODF = \
+        ODFg * (1. + 1/24. * \
+        (kt[0] * (3*U[0, 0]*U[0, 0] - 6*(alpha+1)*U[0, 0]*V00 + \
```

Comment:
This function could be shortened substantially by making this repeated calculations into helper functions. Might improve readability.